### PR TITLE
Datagenned Textures for existing pride robit skins, also added more pride flags

### PR DESCRIPTION
**Pull request by agnor99**
The TextureDatagen uses one method annotated with VisibleForTesting, but it should be fine to keep, for now, as, once this doesn't work, the generated textures can just be moved into the normal resources folder

## Changes proposed in this pull request:
Datageneration of pride robit skins as new pride flags are much simpler to add, just by adding a value to the RobitPrideSkinData enum.

### CODE REVIEW

File: src/datagen/main/java/mekanism/client/texture/PrideRobitTextureProvider.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.server.packs.PackType;
+import net.minecraft.server.packs.resources.Resource;
+import net.minecraftforge.common.data.ExistingFileHelper;
+
+import javax.imageio.ImageIO;
+import java.awt.image.BufferedImage;
+import java.awt.image.ColorModel;
+import java.awt.image.WritableRaster;
+import java.io.*;
```
Comment (pupnewfster) [inline src/datagen/main/java/mekanism/client/texture/PrideRobitTextureProvider.java:18]: We don't use wildcard imports in Mekanism

File: src/datagen/main/java/mekanism/client/texture/PrideRobitTextureProvider.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+import net.minecraft.data.DataProvider;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.server.packs.PackType;
+import net.minecraft.server.packs.resources.Resource;
+import net.minecraftforge.common.data.ExistingFileHelper;
+
+import javax.imageio.ImageIO;
+import java.awt.image.BufferedImage;
+import java.awt.image.ColorModel;
+import java.awt.image.WritableRaster;
```
Comment (pupnewfster) [inline src/datagen/main/java/mekanism/client/texture/PrideRobitTextureProvider.java:17]: iirc Java AWT cannot be run on macs. While I don't use a mac we shouldn't prevent the datagenerator from being able to be ran on a mac.

File: src/main/java/mekanism/common/registries/MekanismRobitSkins.java [hunk 3]
```diff
@@ -13,7 +18,14 @@ private MekanismRobitSkins() {
...
+
+    public static final Map<RobitPrideSkinData, RobitSkinRegistryObject<RobitSkin>> PRIDE_SKINS;
+
+    static {
+        Map<RobitPrideSkinData, RobitSkinRegistryObject<RobitSkin>> tempMap = new HashMap<>();
+        for (RobitPrideSkinData data: RobitPrideSkinData.values()) {
+            tempMap.put(data, ROBIT_SKINS.register(data.lowerCaseName(), data.getColor().length));
+        }
+        PRIDE_SKINS = ImmutableMap.copyOf(tempMap);
+    }
```
Comment (pupnewfster) [inline src/main/java/mekanism/common/registries/MekanismRobitSkins.java:30]: To ensure this maintains order between runs and then gives the registry id it expects I feel like this should instead be:
```java
public static final Map<RobitPrideSkinData, RobitSkinRegistryObject<RobitSkin>> PRIDE_SKINS = Util.make(() -> {
    Map<RobitPrideSkinData, RobitSkinRegistryObject<RobitSkin>> internal = new EnumMap<>(RobitPrideSkinData.class);
    for (RobitPrideSkinData data: RobitPrideSkinData.values()) {
        internal .put(data, ROBIT_SKINS.register(data.lowerCaseName(), data.getColor().length));
    }
    return Collections.unmodifiableMap(internal);
});
```

**Review by pupnewfster — CHANGES_REQUESTED** [review]
I haven't looked too closely at the actual output yet, but left some comments based on a brief skim.

File: src/main/java/mekanism/common/entity/RobitPrideSkinData.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    RobitPrideSkinData(int... color) {
+        this.color = color;
+    }
+
+    public int[] getColor() {
+        return color;
+    }
+
+    public String lowerCaseName() {
+        return name().toLowerCase(Locale.ENGLISH);
```
[the remark this answers is not in the document]
Comment (agnor99) [reply src/main/java/mekanism/common/entity/RobitPrideSkinData.java:26]: I have to disagree, as Locale.ROOT uses the users locale causing Large i's to be conveted to ı, a i without a dot for users with turkish as their system language and this is not a valid character in a registry name.
I'll take a look at the other things tomorrow

File: src/datagen/main/java/mekanism/client/texture/PrideRobitTextureProvider.java (hunk 2 above)
Comment (agnor99) [reply src/datagen/main/java/mekanism/client/texture/PrideRobitTextureProvider.java:17]: I don't think it's possible to do ImageManipulation without awt (except implementing bytereading yourself but I'm not keen on doing that)

File: src/main/java/mekanism/common/entity/RobitPrideSkinData.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (pupnewfster) [reply src/main/java/mekanism/common/entity/RobitPrideSkinData.java:26]: https://garygregory.wordpress.com/2015/11/03/java-lowercase-conversion-turkey/ locale.root is more generic than locale.english and handles cases like you are saying it has issues with.

File: src/datagen/main/java/mekanism/client/texture/PrideRobitTextureProvider.java (hunk 2 above)
Comment (pupnewfster) [reply src/datagen/main/java/mekanism/client/texture/PrideRobitTextureProvider.java:17]: I will look into it some sometime in the next few days, but worst case scenario we could just use a third party library and only load it for datagen.

**Comment by pupnewfster** [comment]
Also remove the changes to the non modified cache files. There is no reason to update them, only reason they are being updated is https://github.com/MinecraftForge/MinecraftForge/pull/8709 hasn't been merged yet.

File: src/datagen/main/java/mekanism/client/texture/PrideRobitTextureProvider.java (hunk 2 above)
Comment (pupnewfster) [reply src/datagen/main/java/mekanism/client/texture/PrideRobitTextureProvider.java:17]: Also I might be remembering wrong and awt may work on macs? Need to look into it further

File: src/main/java/mekanism/common/entity/RobitPrideSkinData.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (agnor99) [reply src/main/java/mekanism/common/entity/RobitPrideSkinData.java:26]: my bad, misremebered the difference between root and default locale

File: src/datagen/main/java/mekanism/client/texture/PrideRobitTextureProvider.java (hunk 2 above)
Comment (pupnewfster) [reply src/datagen/main/java/mekanism/client/texture/PrideRobitTextureProvider.java:17]: Okay I am just dumb, maybe it was awt.Color that isn't available or maybe I am thinking of some other API, but looks like MC does use awt in a variety of places so this should be relatively safe.

**Comment by pupnewfster** [comment]
Once you rebase this to fix the merge conflicts and fix the various comments I made on it I will go ahead and test it locally and then merge it if I have no further comments.

**Review by pupnewfster — CHANGES_REQUESTED** [review]
I went ahead and did some minor formatting cleanup to the PrideRobitTextureProvider and pushed it to your branch, but there are still a few minor issues related to it. I also think for the purposes of looks it should have some logic included so that in cases like with the trans pride robit that I did manually, that given it starts and ends with the same color, that then it should just overlap it so that it doesn't have the weird "double" length bit and looks a bit better.

File: src/main/java/mekanism/common/entity/RobitPrideSkinData.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
+package mekanism.common.entity;
+
+import java.util.Locale;
+
+public enum RobitPrideSkinData {
+    PRIDE(0xFFD12229, 0xFFF68A1E, 0xFFFDE01A, 0xFF007940, 0xFF24408E, 0xFF732982),
+    LESBIAN(0xFFD52D00, 0xFFEA6820, 0xFFFA8F47, 0xFFFDD1B3, 0xFFE9B8D6, 0xFFC85E9E, 0xFFB24687, 0xFFA30262),
+    TRANS(0xFF732982, 0xFFF5A9B8, 0xFFFFFFFF, 0xFFF5A9B8, 0xFF732982),
+    ARO(0xFF3DA745, 0xFFA7D37A, 0xFFFFFFFF, 0xFFABABAB, 0xFF000000),
```
[the remark this answers is not in the document]
Comment (agnor99) [reply src/main/java/mekanism/common/entity/RobitPrideSkinData.java:9]: It is aromatic, but I'm not sure if Aromantic / Asexual / Non Binary written out is the better way, as I thought that their shorter version is as known as the usual long version. 
Also bi and pan didn't get their ending. Aromatic and Asexual are commonly shortened to aro/ace, so I think it's fine?

[the remark this answers is not in the document]
Comment (pupnewfster) [reply src/main/java/mekanism/common/entity/RobitPrideSkinData.java:9]: Yeah sounds fine then, I just wasn't sure about if it was commonly shortened to aro or not, the other ones I was aware of them usually being shortened.

File: src/main/java/mekanism/common/entity/RobitPrideSkinData.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
+package mekanism.common.entity;
+
+import java.util.Locale;
+
+public enum RobitPrideSkinData {
+    PRIDE(0xFFD12229, 0xFFF68A1E, 0xFFFDE01A, 0xFF007940, 0xFF24408E, 0xFF732982),
+    LESBIAN(0xFFD52D00, 0xFFEA6820, 0xFFFA8F47, 0xFFFDD1B3, 0xFFE9B8D6, 0xFFC85E9E, 0xFFB24687, 0xFFA30262),
+    TRANS(0xFF732982, 0xFFF5A9B8, 0xFFFFFFFF, 0xFFF5A9B8, 0xFF732982),
```
[the remark this answers is not in the document]
Comment (agnor99) [reply src/main/java/mekanism/common/entity/RobitPrideSkinData.java:8]: I used a color picker to remake the existing pride robits. I might have made a mistake there.

[the remark this answers is not in the document]
Comment (agnor99) [reply src/main/java/mekanism/common/entity/RobitPrideSkinData.java:8]: I'll check all of them to make sure that there isn't a second problem in there

File: src/main/java/mekanism/common/entity/RobitPrideSkinData.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
+package mekanism.common.entity;
+
+import java.util.Locale;
+
+public enum RobitPrideSkinData {
```
[the remark this answers is not in the document]
Comment (agnor99) [reply src/main/java/mekanism/common/entity/RobitPrideSkinData.java:5]: Will do that, any more wishes?

[the remark this answers is not in the document]
Comment (pupnewfster) [reply src/main/java/mekanism/common/entity/RobitPrideSkinData.java:5]: Nah, I think those should be fine for now.

**Pull request merged**